lib/lambda_ethereum_consensus/state_transition/operations.ex

```diff
@@ -24,92 +22,74 @@ defmodule LambdaEthereumConsensus.StateTransition.Operations do
-              state
-            end
-
-          max_withdrawals_per_payload = ChainSpec.get("MAX_WITHDRAWALS_PER_PAYLOAD")
-
-          # Update the next validator index to start the next withdrawal sweep
-          response =
-            if length(expected_withdrawals) == max_withdrawals_per_payload do
-              # Next sweep starts after the latest withdrawal's validator index
-              latest_withdrawal = List.last(expected_withdrawals)
-              %Withdrawal{validator_index: validator_index} = latest_withdrawal
-              next_validator_index = rem(validator_index + 1, length(validators))
-
-              newer_state = %BeaconState{
-                new_state
-                | next_withdrawal_validator_index: next_validator_index
-              }
-
-              {:ok, newer_state}
-            else
-              # Advance sweep by the max length of the sweep if there was not a full set of withdrawals
-              max_validators_per_withdrawals_sweep =
-                ChainSpec.get("MAX_VALIDATORS_PER_WITHDRAWALS_SWEEP")
 
-              next_index = next_withdrawal_validator_index + max_validators_per_withdrawals_sweep
-              next_validator_index = rem(next_index, length(validators))
+    length_of_expected_withdrawals = length(expected_withdrawals)
 
-              newer_state = %BeaconState{
-                new_state
-                | next_withdrawal_validator_index: next_validator_index
-              }
+    with {:ok, state} <- decrease_balances(state, withdrawals, expected_withdrawals) do
+      # Update the next withdrawal index if this block contained withdrawals
+      new_state =
+        case length_of_expected_withdrawals != 0 do
+          true ->
+            latest_withdrawal = List.last(expected_withdrawals)
+            state |> update_next_withdrawal_index(latest_withdrawal.index + 1)
 
-              {:ok, newer_state}
-            end
-
-          response
+          false ->
+            state
+        end
 
-        {:error, reason} ->
-          {:error, reason}
+      case length_of_expected_withdrawals == ChainSpec.get("MAX_WITHDRAWALS_PER_PAYLOAD") do
+        # Update the next validator index to start the next withdrawal sweep
+        true ->
+          latest_withdrawal = List.last(expected_withdrawals)
+          next_validator_index = rem(latest_withdrawal.validator_index + 1, length(validators))
+          newer_state = new_state |> update_next_withdrawal_validator_index(next_validator_index)
+          {:ok, newer_state}
+
+        # Advance sweep by the max length of the sweep if there was not a full set of withdrawals
+        false ->
+          next_index =
+            next_withdrawal_validator_index +
+              ChainSpec.get("MAX_VALIDATORS_PER_WITHDRAWALS_SWEEP")
+
+          next_validator_index = rem(next_index, length(validators))
+          newer_state = new_state |> update_next_withdrawal_validator_index(next_validator_index)
+          {:ok, newer_state}
```

Comment:
I'd change the general structure of this like this to be:

```elixir
with {:ok, state} <- decrease_balances(state, withdrawals, expected_withdrawals) do
  state
  |> update_next_withdrawal_index(length_of_expected_withdrawals)
  |> update_next_withdrawal_validator_index(length_of_expected_withdrawals)
end
```

This way we don't create "new state" and "newer state" variables. The `ifs` would go in the functions too. For instance:

```elixir
def update_next_withdrawal_index(state, length_of_expected_withdrawals) do
  if length_of_expected_withdrawals == ChainSpec.get("MAX_WITHDRAWALS_PER_PAYLOAD") do
    latest_withdrawal = List.last(expected_withdrawals)
    next_validator_index = rem(latest_withdrawal.validator_index + 1, length(validators))
    {:ok, state |> update_next_withdrawal_validator_index(next_validator_index)}
  else
    next_index = next_withdrawal_validator_index + ChainSpec.get("MAX_VALIDATORS_PER_WITHDRAWALS_SWEEP")
    next_validator_index = rem(next_index, length(validators))
    {:ok, new_state |> update_next_withdrawal_validator_index(next_validator_index)}
  end
end 
```

A similar thing can be done with the other function.

Reply:
Makes sense!